statsmodels/tsa/statespace/mlemodel.py

```diff
@@ -0,0 +1,797 @@
+            attribute. Default is True.
+
+        Notes
+        -----
+        [1]_ recommend maximizing the average likelihood to avoid scale issues;
+        this can be achieved by setting `average_loglike=True`.
+
+        References
+        ----------
+        .. [1] Koopman, Siem Jan, Neil Shephard, and Jurgen A. Doornik. 1999.
+           Statistical Algorithms for Models in State Space Using SsfPack 2.2.
+           Econometrics Journal 2 (1): 107-60. doi:10.1111/1368-423X.00023.
+
+        See Also
+        --------
+        update : modifies the internal state of the Model to reflect new params
+        """
+        if params is not None:
+            self.update(params, transformed=transformed, set_params=set_params)
+
+        # By default, we do not need to consider recreating the entire
+        # _statespace and Cython Kalman filter objects because only parameters
+        # will be changing and not dimensions of matrices.
+        kwargs.setdefault('recreate', False)
+
+        loglike = super(MLEModel, self).loglike(*args, **kwargs)
+
+        # Koopman, Shephard, and Doornik recommend maximizing the average
+        # likelihood to avoid scale issues.
+        if average_loglike:
+            return loglike / self.nobs
+        else:
+            return loglike
+
+    def score(self, params, *args, **kwargs):
+        """
+        Compute the score function at params.
+
+        Parameters
+        ----------
+        params : array_like
+            Array of parameters at which to evaluate the score.
+
+        Returns
+        ----------
+        score : array
+            Score, evaluated at `params`.
+
+        Notes
+        -----
+        This is a numerical approximation.
+        """
+        nargs = len(args)
+        if nargs < 1:
+            kwargs.setdefault('average_loglike', True)
+        if nargs < 2:
+            kwargs.setdefault('transformed', False)
+        if nargs < 3:
+            kwargs.setdefault('set_params', False)
+
+        initial_state = kwargs.get('initial_state', None)
+        initial_state_cov = kwargs.get('initial_state_cov', None)
+        if initial_state is not None and initial_state_cov is not None:
+            # If initialization is stationary, we don't want to recalculate the
+            # initial_state_cov for each new set of parameters here
+            initialization = self.initialization
+            _initial_state = self._initial_state
+            _initial_state_cov = self._initial_state_cov
+            _initial_variance = self._initial_variance
+
+            self.initialize_known(initial_state, initial_state_cov)
+
+        score = approx_fprime_cs(params, self.loglike, epsilon=1e-9, args=args,
+                                 kwargs=kwargs)
+
+        if initial_state is not None and initial_state_cov is not None:
+            # Reset the initialization
+            self.initialization = initialization
```

Comment:
`initialization` is always initialized in `Representation.__init__`, but it may be `None`. What this does is allow passing an *arbitrary* initial state for use calculating the score or Hessian without affecting the default initialization for typical filtering or loglikelihood calculation.

The reason this comes up is this:
- If we are at the MLE and we then want to compute the score Hessian at that point, then `approx_cs_fprime` and `approx_cs_hess` do that by altering the parameters a little bit and calculating the loglikelihood, etc.
- However in the case of a stationary model, the `initial_state_cov` would change a little bit given the altered parameters. But given that we have found the MLE, we want to keep the initial state and initial state cov at the values that would be implied by the MLE, not the altered version.
- Practically, allowing the initial state to be recalculated based on the altered parameters induces instability in the Hessian calculation

So all this does is temporarily use the fixed initialization, and then return it to whatever it was before. This is to allow calling `fit()`, then calculating the covariance parameters, and then calling `fit()` again, without messing up the second `fit()`.